src/util/text.ts

```diff
@@ -250,177 +250,404 @@ export function isAlphaOrDecimal(c: i32): bool {
   return isAlpha(c) || isDecimal(c);
 }
 
-/** Tests if the specified character code is a valid start of an identifier. */
-export function isIdentifierStart(c: i32): bool {
-  return isAlpha(c)
-      || c == CharCode._
-      || c == CharCode.DOLLAR
-      || c >= 170 && c <= 65500
-         && lookupInUnicodeMap(c as u16, unicodeIdentifierStart);
+/** Tests if the specified code point is a valid start of an identifier. */
+export function isIdentifierStart(cp: i32): bool {
+  return isAlpha(cp)
+      || cp == CharCode._
+      || cp == CharCode.DOLLAR
+      || cp >= 170 && cp <= unicodeIdentifierStart[unicodeIdentifierStart.length - 1]
+         && lookupInUnicodeMap(cp, unicodeIdentifierStart);
 }
 
-/** Tests if the specified character code is a valid part of an identifier. */
-export function isIdentifierPart(c: i32): bool {
-  return isAlphaOrDecimal(c)
-      || c == CharCode._
-      || c == CharCode.DOLLAR
-      || c >= 170 && c <= 65500
-         && lookupInUnicodeMap(c as u16, unicodeIdentifierPart);
+/** Tests if the specified code point is a valid part of an identifier. */
+export function isIdentifierPart(cp: i32): bool {
+  return isAlphaOrDecimal(cp)
+      || cp == CharCode._
+      || cp == CharCode.DOLLAR
+      || cp >= 170 && cp <= unicodeIdentifierPart[unicodeIdentifierPart.length - 1]
+         && lookupInUnicodeMap(cp, unicodeIdentifierPart);
 }
 
 /** Tests if the specified string is a valid identifer. */
 export function isIdentifier(str: string): bool {
   var len = str.length;
   if (!len) return false;
-  if (!isIdentifierStart(str.charCodeAt(0))) return false;
-  for (let i = 1; i < len; ++i) {
-    if (!isIdentifierPart(str.charCodeAt(i))) return false;
+  var cp = <i32>str.codePointAt(0);
```

Comment:
This is an interesting one. `"".charCodeAt(0)` returns `NaN`, while `"".codePointAt(0)` returns `undefined`, making this necessary when using standard TS typings.

Reply:
How about `str.codePointAt(0) | 0` ?

Reply:
Here, `undefined` is never really returned because we are always accessing in range, so a noop `<i32>` should do. What bothers me is general usage of this API, where users will face the same choice. Somewhat unfortunate, similar to `Map#get`.